Code/GraphMol/MolDraw2D/AtomSymbol.cpp

```diff
@@ -0,0 +1,141 @@
+  if (symbol_.empty()) {
+    return;
+  }
+  textDrawer_.getStringRects(symbol_, orient_, rects_, drawModes_, drawChars_,
+                             false, TextAlignType::MIDDLE);
+  adjustColons();
+}
+
+// ****************************************************************************
+void AtomSymbol::findExtremes(double &xmin, double &xmax, double &ymin,
+                              double &ymax) const {
+  Point2D bl, br, tl, tr, origTrans;
+  for (auto rect : rects_) {
+    origTrans = rect->trans_;
+    rect->trans_ += cds_;
+    rect->calcCorners(tl, tr, br, bl, 0.0);
+    // sometimes the rect is in a coordinate frame where +ve y is down,
+    // sometimes it's up.  For these purposes, we don't care so long as
+    // the ymax is larger than the ymin.  We probably don't need to do
+    // all the tests for xmin and xmax.
+    xmin = std::min({tr.x, bl.x, xmin});
+    ymin = std::min({tr.y, bl.y, ymin});
+    xmax = std::max({tr.x, bl.x, xmax});
+    ymax = std::max({tr.y, bl.y, ymax});
+    rect->trans_ = origTrans;
+  }
+}
+
+// ****************************************************************************
+void AtomSymbol::scale(const Point2D &scaleFactor) {
+  cds_.x *= scaleFactor.x;
+  cds_.y *= scaleFactor.y;
+
+  // rebuild the rectangles, because the fontScale may be different,
+  // and the widths etc might not scale by the same amount.
+  rects_.clear();
+  drawModes_.clear();
+  drawChars_.clear();
+  textDrawer_.getStringRects(symbol_, orient_, rects_, drawModes_, drawChars_,
+                             false, TextAlignType::MIDDLE);
+}
+
+// ****************************************************************************
+void AtomSymbol::move(const Point2D &trans) { cds_ += trans; }
+
+// ****************************************************************************
+void AtomSymbol::draw(MolDraw2D &molDrawer) const {
+  std::string o_class = molDrawer.getActiveClass();
+  std::string actClass = o_class;
+  if (!actClass.empty()) {
+    actClass += " ";
+  }
+  actClass += boost::str(boost::format("atom-%d") % atIdx_);
+  molDrawer.setActiveClass(actClass);
+  textDrawer_.setColour(colour_);
+  textDrawer_.drawString(symbol_, cds_, orient_);
+  molDrawer.setActiveClass(o_class);
+  //  drawRects(molDrawer);
+}
+
+// ****************************************************************************
+bool AtomSymbol::doesRectClash(const StringRect &rect, double padding) const {
+  for (auto &alrect : rects_) {
+    StringRect r{*alrect};
+    r.trans_ += cds_;
+    if (r.doesItIntersect(rect, padding)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// ****************************************************************************
+void AtomSymbol::adjustColons() {
+  if (symbol_.empty()) {
+    return; // but probable it's always got something in it.
+  }
+  size_t colonPos = symbol_.find(':');
+  if (colonPos == std::string::npos) {
+    return;
+  }
+  double leftHeight = colonPos ? rects_[colonPos - 1]->height_ : 0;
```

Comment:
It's ok for there not to be a colon in the string, in fact it will be the commonest case.  Most atom symbols won't have atom maps in them.
There's a different bug, though, exposed by <lit>[CH2;X2:4]</lit> which is that the markup doesn't have a StringRect so the indexing in the string doesn't match the indexing in the StringRects.  Not sure why that didn't show up in testing before.

Reply:
> It's ok for there not to be a colon in the string, in fact it will be the commonest case. Most atom symbols won't have atom maps in them. 

I think you might have misunderstood where I was suggesting to put that test:
```
  if (colonPos == std::string::npos) {
    return;
  }
  CHECK_INVARIANT(colonPos<=rects_.size(),"bad rects_ size");
  double leftHeight = colonPos ? rects_[colonPos - 1]->height_ : 0;
```
it's not super important, but this would avoid the situation where the colon position is somehow beyond the end of your rectangles array.

Reply:
I had understood what you meant, but I was convinced that it could never happen.  And then it did.  I've fixed that occurrence, but I guess it will do no harm to check and exit a bit less gracelessly.  It will be indicative of a bug if it is fired.